README.md
# NAME

RPi::UnicornHatHD - Use Pimoroni's Awesome Unicorn HAT HD in Perl

# SYNOPSIS

        use RPi::UnicornHatHD;
        my $display = RPi::UnicornHatHD->new();
        while (1) { # Mini rave!
                $display->set_all(sprintf '#%06X', int rand(hex 'FFFFFF'));
                for (0 .. 100, reverse 0 .. 100) {
                        $display->brightness($_ / 100);
                        $display->show();
                }
        }

# DESCRIPTION

Pimoroni's Unicorn HAT HD crams 256 RGB LEDs, in a 16x16 matrix, onto a single HAT. Use it for scrolling news or stock quotes. Mount it somewhere as a mood light. Build a binary clock. Uh. I don't know, you'll think of something.

# METHODS

Use these to make pretty pictures.

## `new()`

        $display = RPi::UnicornHatHD->new();

Creates a new object representing your Unicorn Hat HD. Obviously.

## `brightness($b)`

        $display->brightness(.25); # For tinkering at 1a

Set the display brightness between `0.0` and `1.0`. The default is `0.5`.

## `clear()`

        $display->clear;

Clears the display matrix.

This does not clear the display; it simply resets the 'canvas' for you.

## `get_pixel($x, $y)`

        my ($r, $g, $b) = $display->get_pixel(10, 15);

Returns the color this pixel will display.

## `off()`

        $display->off;

Clears the display matrix and immediately updates the Unicorn Hat HD.

This turns off all the pixels.

## `rotation($r)`

        $display->rotation(180);

Set the display rotation in degrees. Actual rotation will be snapped to the
nearest 90 degrees.

## `set_all($r, $g, $b)`

        $display->set_all(0xFF, 0, 0);
        $display->set_all('#FF0000');

Turns the entire display a single color.

Either...

        $r = Amount of red from 0 to 255
        $g = Amount of green from 0 to 255
        $b = Amount of blue from 0 to 255

...or...

        $h = Hex triplet from #000000 to #FFFFFF

## `set_pixel($x, $y, $r, $g, $b)`

        for my $x (1..10) {
                $display->set_pixel($x, 10, 1, 1, 1);
        }
        $display->set_pixel(0, 0, '#FFF000');

Set a single pixel to RGB color.

        $x = Horizontal position from 0 to 15
        $y = Vertical position from 0 to 15

...and either...

        $r = Amount of red from 0 to 255
        $g = Amount of green from 0 to 255
        $b = Amount of blue from 0 to 255

...or...

        $h = Hex triplet from #000000 to #FFFFFF

## `show()`

        $display->show;

Outputs the contents of the matrix buffer to your Unicorn HAT HD.

# SEE ALSO

- Buy one: [http://shop.pimoroni.com/products/unicorn-hat-hd](http://shop.pimoroni.com/products/unicorn-hat-hd)
- GPIO Pinout: [http://pinout.xyz/pinout/unicorn\_hat\_hd](http://pinout.xyz/pinout/unicorn_hat_hd)
- Github: [https://github.com/sanko/RPi-UnicornHatHD](https://github.com/sanko/RPi-UnicornHatHD)

# LICENSE

Copyright (C) Sanko Robinson.

This library is free software; you can redistribute it and/or modify
it under the terms of the Artistic License 2.

# AUTHOR

Sanko Robinson <sanko@cpan.org>


Changes
Revision history for Perl extension RPi-UnicornHatHD

{{$NEXT}}

0.07 2019-07-17T01:12:24Z

	- 'Stable'

0.06 2019-07-16T03:54:32Z

	- Adding demo script
		- ex/tetris.pl - Cheap tetris clone
	- Correcting prereq

0.05 2017-09-30T04:54:53Z

	- Automatically size image for text length in reddit scroll example

0.04 2017-09-29T17:20:29Z

	- Minor fix on vertical alignment of reddit scroll example script

0.03 2017-09-27T02:11:54Z

	- Include correct LICENSE
	- Clean up docs and examples a little

0.02 2017-09-27T00:50:56Z

	- Adding demo scripts
		- ex/reddit_scroll.pl - Display titles from /r/all
		- ex/rave.pl - Flash and fade random colors
		- ex/tardis.pl - Draw a crude time machine on screen

0.01 2017-09-26T23:25:52Z

    - original version



ex/reddit_scroll.pl
use strict;
use warnings;
use RPi::UnicornHatHD;
use Imager;
use HTTP::Tiny;
use JSON::Tiny qw[decode_json];
use Time::HiRes qw[sleep];

# Scroll the front page of Reddit
my $display = RPi::UnicornHatHD->new();
$display->brightness(1);
my $font = Imager::Font->new(
                file => '/usr/share/fonts/truetype/piboto/Piboto-Regular.ttf',
                type => "ft2",
                color => Imager::Color->new("#000033"),
                size  => 12
) or die Imager->errstr;
while (1) {
    my @titles;
    my $response
        = HTTP::Tiny->new->get('https://www.reddit.com/r/all/new.json');
    die "Failed!\n" unless $response->{success};
    my $json = decode_json $response->{content};
    push @titles, map {
        sprintf '[/r/%s] %s (/u/%s)',
            $_->{data}{subreddit}, $_->{data}{title}, $_->{data}{author}
    } @{$json->{data}{children}};
    for my $title (map { $_ . ' --- ' } @titles) {
        my $img = Imager->new(xsize => 1600, ysize => 16);
        $img->box(filled => 1, color => "010101"); # fill the background color
        my ($left, $top, $right, $bottom)
            = $img->align_string(font   => $font,
                                 text   => $title,
                                 x      => 16,
                                 y      => 2,
                                 halign => 'left',
                                 valign => 'top',
                                 aa     => 1
            );
        for my $scroll_position (0 .. $right) {
            for my $x (0 .. 15) {
                for my $y (0 .. 15) {

                    #warn $x + $scroll_position;
                    my $color
                        = $img->getpixel(x => $x + $scroll_position, y => $y);
                    my ($r, $g, $b, $a) = $color->rgba();
                    $display->set_pixel($y, $x, $r, $b, $g);

                    #print "[X:$x, Y:$y] = R:$r, G:$g, B:$b\n";
                }
            }
            $display->show;
            sleep .01;
            $display->clear;
        }
    }
    $display->off;
    sleep 60 * 5;    # Snooze for a few
}


lib/RPi/UnicornHatHD.pm
package RPi::UnicornHatHD;
#
use Moo;
use strictures 2;
use namespace::clean;
#
our $VERSION = "0.06";
#
use WiringPi::API;
#
my $spi_channel      = 0;
my $spi_max_speed_hz = 9000000;
my $sof              = 0x72;

# _DELAY = 1.0/120 # Leave this to the user
#
sub BUILD {
    my ($self, $args) = @_;
    if ((WiringPi::API::spi_setup($spi_channel, $spi_max_speed_hz) < 0)) {
        die "failed to open the SPI bus...\n";
    }
}
has brightness => (is => 'rw', default => sub {.5});
has rotation   => (is => 'rw', default => sub {0});
has matrix => (is => 'lazy', clearer => 'clear');

sub _build_matrix {
    [map {
         [map { [0, 0, 0] } 1 .. 16]
     } 1 .. 16
    ];
}

sub show {
    my @buffer     = $_[0]->_rotate;
    my $brightness = $_[0]->brightness;
    WiringPi::API::spiDataRW(0,
                             [$sof, map { $_ * $brightness } @buffer],
                             1 + scalar @buffer);
}

sub off {
    my $s = shift;
    $s->clear();
    $s->show();
}

sub set_all {
    my $s = shift;
    my ($r, $g, $b)
        = $#_ == 2 ?
        @_
        : (map { hex $_ } $_[0] =~ m[#(\w{2})(\w{2})(\w{2})]);
    for my $x (0 .. 15) {
        for my $y (0 .. 15) {
            $s->matrix->[$x][$y] = [$r, $g, $b];
        }
    }
}

sub set_pixel {
    my $s = shift;
    my ($x, $y, $r, $g, $b)
        = $#_ == 4 ?
        @_
        : (@_[0 .. 1], map { hex $_ } $_[2] =~ m[#(\w{2})(\w{2})(\w{2})]);
    $s->matrix->[$x][$y] = [$r, $g, $b];
}

sub get_pixel {
    my $s = shift;
    my ($x, $y) = @_;
    $s->matrix->[$x][$y];
}

sub _rotate {
    my $s  = shift;
    my $in = $s->matrix;
    for my $rot (0 .. (($s->rotation / 90) % 4 - 1)) {
        my @out  = ();
        my $rows = scalar @$in;
        for my $row (0 .. $rows - 1) {
            $out[$row] = [map { $_->[$row] } reverse @$in];
        }

        #wantarray ? @out : \@out;
        $in = \@out;
    }
    __flatten($in);
}

sub __flatten {
    map { ref $_ ? __flatten(@{$_}) : $_ } @_;
}
1;
__END__

=encoding utf-8

=head1 NAME

RPi::UnicornHatHD - Use Pimoroni's Awesome Unicorn HAT HD in Perl

=head1 SYNOPSIS

	use RPi::UnicornHatHD;
	my $display = RPi::UnicornHatHD->new();
	while (1) { # Mini rave!
		$display->set_all(sprintf '#%06X', int rand(hex 'FFFFFF'));
		for (0 .. 100, reverse 0 .. 100) {
			$display->brightness($_ / 100);
			$display->show();
		}
	}

=head1 DESCRIPTION

Pimoroni's Unicorn HAT HD crams 256 RGB LEDs, in a 16x16 matrix, onto a single HAT for your Raspberry Pi. Use it for scrolling news or stock quotes. Mount it somewhere as a mood light. Build a binary clock. Uh. I don't know, you'll think of something.

=head1 METHODS

Use these to make pretty pictures.

=head2 C<new()>

	$display = RPi::UnicornHatHD->new();

Creates a new object representing your Unicorn Hat HD. Obviously.

=head2 C<brightness($b)>

	$display->brightness(.25); # For tinkering at 1a

Set the display brightness between C<0.0> and C<1.0>. The default is C<0.5>.

=head2 C<clear()>

	$display->clear;

Clears the display matrix.

This does not clear the display; it simply resets the 'canvas' for you.

=head2 C<get_pixel($x, $y)>

	my ($r, $g, $b) = $display->get_pixel(10, 15);

Returns the color this pixel will display.

=head2 C<off()>

	$display->off;

Clears the display matrix and immediately updates the Unicorn Hat HD.

This turns off all the pixels.

=head2 C<rotation($r)>

	$display->rotation(180);

Set the display rotation in degrees. Actual rotation will be snapped to the
nearest 90 degrees.

=head2 C<set_all($r, $g, $b)>

	$display->set_all(0xFF, 0, 0);
	$display->set_all('#FF0000');

Turns the entire display a single color.

Either...

	$r = Amount of red from 0 to 255
	$g = Amount of green from 0 to 255
	$b = Amount of blue from 0 to 255

...or...

	$h = Hex triplet from #000000 to #FFFFFF

=head2 C<set_pixel($x, $y, $r, $g, $b)>

	for my $x (1..10) {
		$display->set_pixel($x, 10, 1, 1, 1);
	}
	$display->set_pixel(0, 0, '#FFF000');

Set a single pixel to RGB color.

	$x = Horizontal position from 0 to 15
	$y = Vertical position from 0 to 15

...and either...

	$r = Amount of red from 0 to 255
	$g = Amount of green from 0 to 255
	$b = Amount of blue from 0 to 255

...or...

	$h = Hex triplet from #000000 to #FFFFFF

=head2 C<show()>

	$display->show;

Outputs the contents of the matrix buffer to your Unicorn HAT HD.

=head1 SEE ALSO

=over

=item * Buy one: L<http://shop.pimoroni.com/products/unicorn-hat-hd>

=item * GPIO Pinout: L<http://pinout.xyz/pinout/unicorn_hat_hd>

=item * Github: L<https://github.com/sanko/RPi-UnicornHatHD>

=back

=head1 LICENSE

Copyright (C) Sanko Robinson.

This library is free software; you can redistribute it and/or modify
it under the terms of the Artistic License 2.

=head1 AUTHOR

Sanko Robinson E<lt>sanko@cpan.orgE<gt>

=cut
